File: packages/anvil/core/src/architecture/baseline.ts
/**
 * Architecture baseline storage
 *
 * Handles reading/writing .anvil/architecture.json
 */

import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { join, dirname } from 'node:path';
import {
  ArchitectureBaselineSchema,
  type ArchitectureBaseline,
  type Layers,
  type EntryPoint,
  type Boundary,
  type BaselineViolation,
  createDefaultLayers,
  createDefaultBoundaries,
} from './types.js';
import { createDebugger } from '../utils/debug.js';

const debug = createDebugger('architecture');

/**
 * Default baseline file path
 */
export const BASELINE_FILENAME = 'architecture.json';
export const ANVIL_DIR = '.anvil';

/**
 * Get the full path to the baseline file
 */
export function getBaselinePath(workspaceRoot: string): string {
  return join(workspaceRoot, ANVIL_DIR, BASELINE_FILENAME);
}

/**
 * Check if a baseline exists
 */
export function baselineExists(workspaceRoot: string): boolean {
  return existsSync(getBaselinePath(workspaceRoot));
}

/**
 * Load the architecture baseline
 */
export function loadBaseline(workspaceRoot: string): ArchitectureBaseline | null {
  const path = getBaselinePath(workspaceRoot);

  if (!existsSync(path)) {
    debug('no baseline file found', path);
    return null;
  }

  try {
    debug('loading baseline from', path);
    const content = readFileSync(path, 'utf-8');
    const data = JSON.parse(content);

    // Validate against schema
    const result = ArchitectureBaselineSchema.safeParse(data);

    if (!result.success) {
      debug('invalid baseline schema', result.error.format());
      console.error('Invalid architecture baseline:', result.error.format());
      return null;
    }

    debug('baseline loaded', {
      modules: result.data.baseline_snapshot.module_count,
      violations: result.data.baseline_snapshot.violations.length,
    });
    return result.data;
  } catch (error) {
    debug('failed to load baseline', error instanceof Error ? error : undefined);
    console.error('Failed to load architecture baseline:', error);
    return null;
  }
}

/**
 * Save the architecture baseline
 */
export function saveBaseline(workspaceRoot: string, baseline: ArchitectureBaseline): void {
  debug('saving baseline', { modules: baseline.baseline_snapshot.module_count });
  const path = getBaselinePath(workspaceRoot);
  const dir = dirname(path);

  // Ensure directory exists
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }

  // Validate before saving
  const result = ArchitectureBaselineSchema.safeParse(baseline);
  if (!result.success) {
    throw new Error(`Invalid baseline data: ${result.error.format()}`);
  }

  // Write with pretty formatting for human readability
  writeFileSync(path, JSON.stringify(baseline, null, 2) + '\n', 'utf-8');
}

/**
 * Create a new baseline with defaults
 */
export function createBaseline(options: {
  entryPoints?: EntryPoint[];
  layers?: Layers;
  boundaries?: Boundary[];
  violations?: BaselineViolation[];
  moduleCount?: number;
}): ArchitectureBaseline {
  const now = new Date().toISOString();
  const layers = options.layers ?? createDefaultLayers();
  const boundaries = options.boundaries ?? createDefaultBoundaries(layers);

  return {
    schema_version: '0.1.0',
    created_at: now,
    updated_at: now,
    entry_points: options.entryPoints ?? [],
    layers,
    boundaries,
    baseline_snapshot: {
      module_count: options.moduleCount ?? 0,
      timestamp: now,
      violations: options.violations ?? [],
    },
  };
}

/**
 * Update an existing baseline
 */
export function updateBaseline(
  existing: ArchitectureBaseline,
  updates: Partial<{
    entryPoints: EntryPoint[];
    layers: Layers;
    boundaries: Boundary[];
    violations: BaselineViolation[];
    moduleCount: number;
  }>
): ArchitectureBaseline {
  const now = new Date().toISOString();

  return {
    ...existing,
    updated_at: now,
    entry_points: updates.entryPoints ?? existing.entry_points,
    layers: updates.layers ?? existing.layers,
    boundaries: updates.boundaries ?? existing.boundaries,
    baseline_snapshot: {
      ...existing.baseline_snapshot,
      module_count: updates.moduleCount ?? existing.baseline_snapshot.module_count,
      violations: updates.violations ?? existing.baseline_snapshot.violations,
      timestamp: now,
    },
  };
}

/**
 * Merge new violations into baseline (for incremental updates)
 */
export function mergeViolations(
  existing: BaselineViolation[],
  newViolations: BaselineViolation[]
): BaselineViolation[] {
  const byId = new Map<string, BaselineViolation>();

  // Add existing
  for (const v of existing) {
    byId.set(v.id, v);
  }

  // Add/update new
  for (const v of newViolations) {
    byId.set(v.id, v);
  }

  return Array.from(byId.values());
}

/**
 * Find violations that are NEW (not in baseline)
 */
export function findNewViolations(
  current: BaselineViolation[],
  baseline: BaselineViolation[]
): BaselineViolation[] {
  const baselineIds = new Set(baseline.map((v) => v.id));
  return current.filter((v) => !baselineIds.has(v.id));
}

/**
 * Find violations that were FIXED (in baseline but not current)
 */
export function findFixedViolations(
  current: BaselineViolation[],
  baseline: BaselineViolation[]
): BaselineViolation[] {
  const currentIds = new Set(current.map((v) => v.id));
  return baseline.filter((v) => !currentIds.has(v.id));
}

/**
 * Baseline manager for convenient operations
 */
export class BaselineManager {
  private workspaceRoot: string;
  private baseline: ArchitectureBaseline | null = null;

  constructor(workspaceRoot: string) {
    this.workspaceRoot = workspaceRoot;
  }

  /**
   * Check if baseline exists
   */
  exists(): boolean {
    return baselineExists(this.workspaceRoot);
  }

  /**
   * Load baseline (cached)
   */
  load(): ArchitectureBaseline | null {
    if (!this.baseline) {
      this.baseline = loadBaseline(this.workspaceRoot);
    }
    return this.baseline;
  }

  /**
   * Force reload baseline
   */
  reload(): ArchitectureBaseline | null {
    this.baseline = loadBaseline(this.workspaceRoot);
    return this.baseline;
  }

  /**
   * Save baseline
   */
  save(baseline: ArchitectureBaseline): void {
    saveBaseline(this.workspaceRoot, baseline);
    this.baseline = baseline;
  }

  /**
   * Create and save a new baseline
   */
  create(options: Parameters<typeof createBaseline>[0]): ArchitectureBaseline {
    const baseline = createBaseline(options);
    this.save(baseline);
    return baseline;
  }

  /**
   * Update and save existing baseline
   */
  update(updates: Parameters<typeof updateBaseline>[1]): ArchitectureBaseline | null {
    const existing = this.load();
    if (!existing) {
      return null;
    }

    const updated = updateBaseline(existing, updates);
    this.save(updated);
    return updated;
  }

  /**
   * Get layers from baseline or defaults
   */
  getLayers(): Layers {
    const baseline = this.load();
    return baseline?.layers ?? createDefaultLayers();
  }

  /**
   * Get boundaries from baseline or defaults
   */
  getBoundaries(): Boundary[] {
    const baseline = this.load();
    if (baseline) {
      return baseline.boundaries;
    }
    return createDefaultBoundaries(createDefaultLayers());
  }

  /**
   * Check if a violation is new
   */
  isNewViolation(violation: BaselineViolation): boolean {
    const baseline = this.load();
    if (!baseline) {
      return true; // No baseline = all violations are new
    }

    return !baseline.baseline_snapshot.violations.some((v) => v.id === violation.id);
  }

  /**
   * Get baseline path
   */
  getPath(): string {
    return getBaselinePath(this.workspaceRoot);
  }
}

/**
 * Create a baseline manager
 */
export function createBaselineManager(workspaceRoot: string): BaselineManager {
  return new BaselineManager(workspaceRoot);
}

File: crates/anvil-architecture/src/baseline.rs
// Baseline management — load, save, create `.anvil/architecture.json`.

use std::collections::BTreeMap;
use std::path::Path;

use chrono::Utc;

use crate::ANVIL_DIR;
use crate::types::{
    ArchitectureBaseline, BaselineSnapshot, BaselineViolation, Boundary, EntryPoint, Layers,
    create_default_boundaries, create_default_layers,
};
use crate::util::atomic_write;

/// Baseline file name.
pub const BASELINE_FILENAME: &str = "architecture.json";

/// Errors from baseline operations.
#[derive(Debug, thiserror::Error)]
pub enum BaselineError {
    #[error("I/O error at {path}: {source}")]
    Io {
        path: String,
        #[source]
        source: std::io::Error,
    },
    #[error("invalid baseline JSON: {0}")]
    InvalidJson(String),
}

/// Get the full path to the baseline file.
pub fn get_baseline_path(workspace_root: &Path) -> std::path::PathBuf {
    workspace_root.join(ANVIL_DIR).join(BASELINE_FILENAME)
}

/// Check whether a baseline file exists.
pub fn baseline_exists(workspace_root: &Path) -> bool {
    get_baseline_path(workspace_root).exists()
}

/// Load the architecture baseline from `.anvil/architecture.json`.
///
/// Returns `Ok(None)` if the file does not exist.
pub fn load_baseline(workspace_root: &Path) -> Result<Option<ArchitectureBaseline>, BaselineError> {
    load_baseline_capped(workspace_root, MAX_BASELINE_BYTES)
}

/// Maximum size of `.anvil/architecture.json` that [`load_baseline`] reads into
/// memory (CIB-084). A baseline larger than this is almost certainly corrupt or
/// hostile; the read is refused rather than committing unbounded memory.
pub const MAX_BASELINE_BYTES: u64 = 16 * 1024 * 1024;

/// [`load_baseline`] with an explicit read cap, so the size guard is testable
/// without writing a multi-megabyte fixture.
fn load_baseline_capped(
    workspace_root: &Path,
    cap: u64,
) -> Result<Option<ArchitectureBaseline>, BaselineError> {
    let path = get_baseline_path(workspace_root);

    if !path.exists() {
        return Ok(None);
    }

    let path_str = path.display().to_string();
    let content =
        crate::util::read_to_string_capped(&path, cap).map_err(|e| BaselineError::Io {
            path: path_str,
            source: e,
        })?;

    let baseline: ArchitectureBaseline =
        serde_json::from_str(&content).map_err(|e| BaselineError::InvalidJson(e.to_string()))?;

    Ok(Some(baseline))
}

/// Save the architecture baseline to `.anvil/architecture.json`.
pub fn save_baseline(
    workspace_root: &Path,
    baseline: &ArchitectureBaseline,
) -> Result<(), BaselineError> {
    let path = get_baseline_path(workspace_root);
    let path_str = path.display().to_string();

    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent).map_err(|e| BaselineError::Io {
            path: path_str.clone(),
            source: e,
        })?;
    }

    let content = serde_json::to_string_pretty(baseline)
        .map_err(|e| BaselineError::InvalidJson(e.to_string()))?;

    atomic_write(&path, format!("{content}\n").as_bytes()).map_err(|e| BaselineError::Io {
        path: path_str,
        source: e,
    })?;

    Ok(())
}

/// Options for creating a baseline.
#[derive(Debug, Default)]
pub struct CreateBaselineOptions {
    pub entry_points: Vec<EntryPoint>,
    pub layers: Option<Layers>,
    pub boundaries: Option<Vec<Boundary>>,
    pub violations: Vec<BaselineViolation>,
    pub module_count: u32,
}

/// Create a new baseline with defaults.
pub fn create_baseline(options: CreateBaselineOptions) -> ArchitectureBaseline {
    let now = Utc::now().to_rfc3339();
    let layers = options.layers.unwrap_or_else(create_default_layers);
    let boundaries = options
        .boundaries
        .unwrap_or_else(|| create_default_boundaries(&layers));

    ArchitectureBaseline {
        schema_version: "0.1.0".into(),
        created_at: now.clone(),
        updated_at: now.clone(),
        entry_points: options.entry_points,
        layers,
        boundaries,
        baseline_snapshot: BaselineSnapshot {
            module_count: options.module_count,
            timestamp: now,
            violations: options.violations,
        },
    }
}

/// Merge new violations into an existing set (deduplicates by ID).
pub fn merge_violations(
    existing: &[BaselineViolation],
    new_violations: &[BaselineViolation],
) -> Vec<BaselineViolation> {
    let mut by_id: BTreeMap<&str, &BaselineViolation> = BTreeMap::new();

    for v in existing {
        by_id.insert(&v.id, v);
    }
    for v in new_violations {
        by_id.insert(&v.id, v);
    }

    by_id.into_values().cloned().collect()
}

/// Find violations that are NEW (not present in baseline).
pub fn find_new_violations(
    current: &[BaselineViolation],
    baseline: &[BaselineViolation],
) -> Vec<BaselineViolation> {
    let baseline_ids: std::collections::HashSet<&str> =
        baseline.iter().map(|v| v.id.as_str()).collect();
    let mut result: Vec<BaselineViolation> = current
        .iter()
        .filter(|v| !baseline_ids.contains(v.id.as_str()))
        .cloned()
        .collect();
    result.sort_by(|a, b| a.id.cmp(&b.id));
    result
}

/// Find violations that were FIXED (in baseline but not in current).
pub fn find_fixed_violations(
    current: &[BaselineViolation],
    baseline: &[BaselineViolation],
) -> Vec<BaselineViolation> {
    let current_ids: std::collections::HashSet<&str> =
        current.iter().map(|v| v.id.as_str()).collect();
    let mut result: Vec<BaselineViolation> = baseline
        .iter()
        .filter(|v| !current_ids.contains(v.id.as_str()))
        .cloned()
        .collect();
    result.sort_by(|a, b| a.id.cmp(&b.id));
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn baseline_exists_returns_false_for_missing() {
        let tmp = tempfile::TempDir::new().unwrap();
        assert!(!baseline_exists(tmp.path()));
    }

    #[test]
    fn load_baseline_rejects_a_file_over_the_read_cap() {
        // CIB-084: an over-cap architecture.json is refused (as an IO error)
        // before it is read into memory or parsed.
        let tmp = tempfile::TempDir::new().unwrap();
        std::fs::create_dir_all(tmp.path().join(ANVIL_DIR)).unwrap();
        std::fs::write(get_baseline_path(tmp.path()), "{}").unwrap();
        let err = load_baseline_capped(tmp.path(), 1).unwrap_err();
        assert!(matches!(err, BaselineError::Io { .. }), "{err:?}");
    }

    #[test]
    fn load_baseline_returns_none_when_missing() {
        let tmp = tempfile::TempDir::new().unwrap();
        let result = load_baseline(tmp.path()).unwrap();
        assert!(result.is_none());
    }

    #[test]
    fn create_and_load_baseline_round_trips() {
        let tmp = tempfile::TempDir::new().unwrap();

        let baseline = create_baseline(CreateBaselineOptions {
            module_count: 42,
            ..Default::default()
        });
        assert_eq!(baseline.schema_version, "0.1.0");
        assert_eq!(baseline.baseline_snapshot.module_count, 42);

        save_baseline(tmp.path(), &baseline).unwrap();
        assert!(baseline_exists(tmp.path()));

        let loaded = load_baseline(tmp.path()).unwrap().unwrap();
        assert_eq!(loaded.schema_version, "0.1.0");
        assert_eq!(loaded.baseline_snapshot.module_count, 42);
    }

    #[test]
    fn load_baseline_rejects_invalid_json() {
        let tmp = tempfile::TempDir::new().unwrap();
        let anvil_dir = tmp.path().join(ANVIL_DIR);
        std::fs::create_dir_all(&anvil_dir).unwrap();
        std::fs::write(anvil_dir.join(BASELINE_FILENAME), "not-json").unwrap();

        let result = load_baseline(tmp.path());
        assert!(matches!(result, Err(BaselineError::InvalidJson(_))));
    }

    #[test]
    fn create_baseline_uses_defaults() {
        let baseline = create_baseline(CreateBaselineOptions::default());
        assert!(!baseline.layers.is_empty());
        assert!(!baseline.boundaries.is_empty());
    }

    fn make_violation(id: &str) -> BaselineViolation {
        BaselineViolation {
            id: id.into(),
            from_layer: "a".into(),
            to_layer: "b".into(),
            from_file: "a.ts".into(),
            to_file: "b.ts".into(),
            import_line: 1,
            rule: None,
        }
    }

    #[test]
    fn merge_violations_deduplicates_and_sorts_by_id() {
        let existing = vec![make_violation("v2"), make_violation("v1")];
        let new = vec![make_violation("v3"), make_violation("v2")];

        let merged = merge_violations(&existing, &new);
        assert_eq!(merged.len(), 3);
        assert_eq!(merged[0].id, "v1");
        assert_eq!(merged[1].id, "v2");
        assert_eq!(merged[2].id, "v3");
    }

    // EATEST-022: When `merge_violations` sees two entries sharing the same
    // ID but different fields (e.g. different `from_file`), the new-list
    // entry must win. Pins the precedence semantics callers depend on when
    // refreshing a baseline.
    #[test]
    fn merge_violations_new_wins_on_id_collision() {
        let existing = vec![BaselineViolation {
            id: "shared".into(),
            from_layer: "a".into(),
            to_layer: "b".into(),
            from_file: "old_path.ts".into(),
            to_file: "b.ts".into(),
            import_line: 1,
            rule: None,
        }];
        let new = vec![BaselineViolation {
            id: "shared".into(),
            from_layer: "a".into(),
            to_layer: "b".into(),
            from_file: "new_path.ts".into(),
            to_file: "b.ts".into(),
            import_line: 9,
            rule: Some("refreshed".into()),
        }];

        let merged = merge_violations(&existing, &new);
        assert_eq!(merged.len(), 1, "ID collision must produce a single entry");
        assert_eq!(
            merged[0].from_file, "new_path.ts",
            "new_violations entry must win on ID collision"
        );
        assert_eq!(merged[0].import_line, 9);
        assert_eq!(merged[0].rule.as_deref(), Some("refreshed"));
    }

    #[test]
    fn find_new_violations_filters_baseline() {
        let current = vec![make_violation("v1"), make_violation("v2")];
        let baseline = vec![make_violation("v1")];

        let new = find_new_violations(&current, &baseline);
        assert_eq!(new.len(), 1);
        assert_eq!(new[0].id, "v2");
    }

    #[test]
    fn find_fixed_violations_filters_current() {
        let current = vec![make_violation("v1")];
        let baseline = vec![make_violation("v1"), make_violation("v2")];

        let fixed = find_fixed_violations(&current, &baseline);
        assert_eq!(fixed.len(), 1);
        assert_eq!(fixed[0].id, "v2");
    }
}
